### PR TITLE
Add an operator bool() to AutoPtr.

### PR DESCRIPTION
**Pull request by jwpeterson**
I don't think this changes the inner workings of AutoPtr in any way,
and allows us to be more compatible with newer smart pointer types.

This fixes the compile error we encountered in #694, so presumably once this is merged we can invalidate that one and it will make it a bit further...

### CODE REVIEW

**Comment by roystgnr** [comment]
:+1: and :100: only because I can't find a :+infinity: emoticon. Thanks for fixing this so quickly.

File: include/base/auto_ptr.h [hunk 1]
```diff
@@ -373,6 +373,13 @@ class AutoPtr
   }
 
   /**
+   * Forward compatibility with newer smart pointer types.  This
+   * allows code like if (!foo) to work with AutoPtr.
+   */
+  explicit operator bool() const
+  { return (this->get() != NULL); }
```
Comment (roystgnr) [inline include/base/auto_ptr.h]: You guys and your "!= NULL" pointer comparisons.  How about we return "(this->get() != NULL) == true" for good measure?  ;-)

**Comment by jwpeterson** [comment]
This actually produces a warning in C++03 code, so I'll guard the use of explicit appropriately before merging...

File: include/base/auto_ptr.h (hunk 1 above)
Comment (jwpeterson) [inline include/base/auto_ptr.h]:
```
if (!this->get())
  that->get()
```

**Comment by jwpeterson** [comment]
The only failures were the tests that time out, so I'm going to merge.

**Pull request merged**